### PR TITLE
obs-studio: 27.0.1 -> 27.2.0

### PR DESCRIPTION
**Pull request by blitz**
###### Motivation for this change

This updates OBS Studio and its companion libcef library to the latest versions. ~~I've failed to find a libcef version that makes OBS Studio 21.1.3 build, so I jumped ahead to the recent beta. We might want to hold off merging this until they have at least reached a RC or something.~~ 27.2.0 is now out!

Testing done:
 - Basic video compositing with pipewire screen capture and USB webcam.

Supersedes #147737.

###### Steps to un-Draft

- [x] Wait for at least an OBS Studio release candidate.
- [ ] ~~Wait for obs-websocket to have a release that works with OBS 27.2 **or** backport the build fixes to 4.9.1.~~
 - This doesn't seem feasible. I've updated obs-websocket to the latest alpha version instead.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/video/obs-studio/default.nix [hunk 1]
```diff
@@ -59,7 +61,7 @@ mkDerivation rec {
     ./Enable-file-access-and-universal-access-for-file-URL.patch
 
     # Lets obs-browser build against CEF 91.1.0+
-    ./Change-product_version-to-user_agent_product.patch
+    # ./Change-product_version-to-user_agent_product.patch
```
Comment (pasqui23) [inline pkgs/applications/video/obs-studio/default.nix:64]: Once it runs, you should remove those lines and the file from nixpkgs, as the patch was incorporated upstream

Comment (blitz) [reply pkgs/applications/video/obs-studio/default.nix:64]: It does run. Forgot about removing this patch. Fixed. :)

**Comment by blitz** [comment]
@ofborg build obs-studio

**Review by dasj19 — COMMENTED** [review]
I agree with you that we should wait for a more stable release of obs-studio.
But the cerf binary update could be released in a separate PR.

**Comment by NilsIrl** [comment]
Result of `nixpkgs-review pr 152697` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>obs-studio-plugins.obs-ndi</li>
    <li>obs-studio-plugins.obs-nvfbc</li>
    <li>obs-studio-plugins.obs-websocket</li>
  </ul>
</details>
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>libcef</li>
    <li>obs-studio</li>
    <li>obs-studio-plugins.looking-glass-obs</li>
    <li>obs-studio-plugins.obs-gstreamer</li>
    <li>obs-studio-plugins.obs-move-transition</li>
    <li>obs-studio-plugins.obs-multi-rtmp</li>
    <li>obs-studio-plugins.wlrobs</li>
  </ul>
</details>

**Comment by blitz** [comment]
@NilsIrl Thanks! I'll look into the failures.

**Comment by blitz** [comment]
@dasj19 Updating libcef separately doesn't work, because the old obs doesn't build against the new version of libcef.

**Comment by NilsIrl** [comment]
```
building '/nix/store/ddk2dzzailyx2cs7w7db617w3v5ggpyx-obs-nvfbc-0.0.4.drv'...
building '/nix/store/7ja58yjrf96873q96zjgnw9hgd5cyrxc-obs-websocket-4.9.1.drv'...
building '/nix/store/691a8if39wpd6r6iwl7bs09kpjw804h0-InstallNDISDK_v4_Linux.tar.gz.drv'...
builder for '/nix/store/691a8if39wpd6r6iwl7bs09kpjw804h0-InstallNDISDK_v4_Linux.tar.gz.drv' failed with exit code 1; last 10 log lines:

    https://ndi.tv/sdk/

  Once you have downloaded the file, please use the following command and
  re-run the installation:

    $ nix-prefetch-url file://$PWD/InstallNDISDK_v4_Linux.tar.gz

  ***
cannot build derivation '/nix/store/cw216jgmk11f13v83w3w4v54q35xr8ww-ndi-4.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/bxqjzfms7kfmdkxhbz5ryfndwaqy4baw-obs-ndi-4.9.1.drv': 1 dependencies couldn't be built
builder for '/nix/store/ddk2dzzailyx2cs7w7db617w3v5ggpyx-obs-nvfbc-0.0.4.drv' failed with exit code 1; last 10 log lines:
  FAILED: nvfbc.so.p/nvfbc.c.o
  gcc -Invfbc.so.p -I. -I.. -I/nix/store/1fbrwmybda87v9cljcl65pnrwcjbm8c3-libglvnd-1.3.4-dev/include -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=c99 -fPIC -pthread -MD -MQ nvfbc.so.p/nvfbc.c.o -MF nvfbc.so.p/nvfbc.c.o.d -o nvfbc.so.p/nvfbc.c.o -c ../nvfbc.c
  In file included from ../nvfbc.c:32:
  /nix/store/aflfv02rhr4hs0ni6b5lqaqjhs9pfbn7-obs-studio-27.2.0-beta1/include/obs/util/threading.h: In function 'pthread_mutex_init_recursive':
  /nix/store/aflfv02rhr4hs0ni6b5lqaqjhs9pfbn7-obs-studio-27.2.0-beta1/include/obs/util/threading.h:61:42: error: 'PTHREAD_MUTEX_RECURSIVE' undeclared (first use in this function); did you mean 'PTHREAD_MUTEX_RECURSIVE_NP'?
     61 |   ret = pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE);
        |                                          ^~~~~~~~~~~~~~~~~~~~~~~
        |                                          PTHREAD_MUTEX_RECURSIVE_NP
  /nix/store/aflfv02rhr4hs0ni6b5lqaqjhs9pfbn7-obs-studio-27.2.0-beta1/include/obs/util/threading.h:61:42: note: each undeclared identifier is reported only once for each function it appears in
  ninja: build stopped: subcommand failed.
builder for '/nix/store/7ja58yjrf96873q96zjgnw9hgd5cyrxc-obs-websocket-4.9.1.drv' failed with exit code 2; last 10 log lines:
                   from /nix/store/ihwd7nsmdwhrpz9h6qsxlaaqpj9dp545-qtbase-5.15.3-dev/include/QtCore/qfuture.h:45,
                   from /nix/store/ihwd7nsmdwhrpz9h6qsxlaaqpj9dp545-qtbase-5.15.3-dev/include/QtCore/QtCore:97,
                   from /nix/store/ihwd7nsmdwhrpz9h6qsxlaaqpj9dp545-qtbase-5.15.3-dev/include/QtConcurrent/QtConcurrentDepends:3,
                   from /nix/store/ihwd7nsmdwhrpz9h6qsxlaaqpj9dp545-qtbase-5.15.3-dev/include/QtConcurrent/QtConcurrent:3,
                   from /build/source/src/WSServer.cpp:26:
  /nix/store/ihwd7nsmdwhrpz9h6qsxlaaqpj9dp545-qtbase-5.15.3-dev/include/QtCore/qmutex.h:140:14: note: declared here
    140 |     explicit QMutex(RecursionMode mode);
        |              ^~~~~~
  make[1]: *** [CMakeFiles/Makefile2:84: CMakeFiles/obs-websocket.dir/all] Error 2
  make: *** [Makefile:136: all] Error 2
```

**Comment by blitz** [comment]
@NilsIrl The obs-studio-plugins.obs-ndi plugin requires file that cannot be freely downloaded. I also don't finde the 4.6.2 version currently required. I only see version 5. So we could consider updating this to the latest version after finishing the OBS update.

**Comment by blitz** [comment]
obs-nvfbc just worked after updating it.

obs-websocket is a bit more problematic, because the patches to make it build with the new OBS Studio are hard to backport to their latest release. The do have an alpha version out that does build against the new OBS, but according to their docs, this has no supported clients yet, so users will find that pretty useless. I've updated it nevertheless for now.

**Comment by blitz** [comment]
@ofborg build obs-studio-plugins.obs-websocket

**Review by lf- — APPROVED** [review]
Confirmed working on aarch64-linux.

**Comment by blitz** [comment]
I've rebased and dropped the obs-nvfbc updated due to #153177.

**Comment by blitz** [comment]
I've updted the PR to OBS Studio 27.2 RC4. I would appreciate if anyone gives this a spin to see if they spot any issues.

**Comment by jansol** [comment]
Would be good to get this into the release channels before the end of the month: https://twitter.com/OBSProject/status/1493351348997398530

**Comment by blitz** [comment]
> Would be good to get this into the release channels before the end of the month: https://twitter.com/OBSProject/status/1493351348997398530

Good to know!

**Comment by blitz** [comment]
I've updated this PR to the new 27.2 release. From my point this is now ready to be merged.

There is one :hankey: in here and that is the alpha version of the obs-websocket plugin. But from where I'm standing, it's either that or marking obs-websocket as broken until they have a release that's compatible with OBS 27.2.

**Comment by blitz** [comment]
obs-studio-plugins.obs-websocket breaks with the current OBS release. Will look into that later. Given that their 5.0-alpha has no know working clients according to their [README](https://github.com/obsproject/obs-websocket#client-software), I'm very tempted to mark it broken until they have a working version.

**Comment by erdnaxe** [comment]
> obs-studio-plugins.obs-websocket breaks with the current OBS release. Will look into that later. Given that their 5.0-alpha has no know working clients according to their [README](https://github.com/obsproject/obs-websocket#client-software), I'm very tempted to mark it broken until they have a working version.

I agree, I believe it is better to ship latest OBS.

If we mark `obs-studio-plugins.obs-websocket` broken, should we include an indication about how to use previous OBS release?

**Review by fpletz — APPROVED** [review]
LGTM. When you're ready with obs-websocket I will gladly merge this.

**Comment by blitz** [comment]
> LGTM. When you're ready with obs-websocket I will gladly merge this.

@fpletz I've pulled in the compilation fix for obs-websocket (see last commit). From my end this is good to go.

**Review by IvarWithoutBones — APPROVED** [review]
Looks good to me :blush: 

The build failure is from `requireFile`, so unrelated

Result of `nixpkgs-review pr 152697` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>obs-studio-plugins.obs-ndi</li>
  </ul>
</details>
<details>
  <summary>9 packages built:</summary>
  <ul>
    <li>libcef</li>
    <li>obs-studio</li>
    <li>obs-studio-plugins.looking-glass-obs</li>
    <li>obs-studio-plugins.obs-gstreamer</li>
    <li>obs-studio-plugins.obs-move-transition</li>
    <li>obs-studio-plugins.obs-multi-rtmp</li>
    <li>obs-studio-plugins.obs-nvfbc</li>
    <li>obs-studio-plugins.obs-websocket</li>
    <li>obs-studio-plugins.wlrobs</li>
  </ul>
</details>

**Comment by fpletz** [comment]
Thanks!

**Pull request merged**